### PR TITLE
Fetch images when previewing Twitter

### PR DESCRIPTION
**Pull request by AndrewRyanChama**
We found that some websites return opengraph information based
on the user agent.

In order to address this, using (bot) in the user agent string
may help the website behave correctly. When testing Twitter, we
found that the correct metadata is returned with the (bot) user
agent while it isn't for the default user agent.

This change additionally updates providers.json, as with the
user agent fix twitter will now return the expected openGraph
metadata.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Before:
Screen Shot 2022-02-13 at 7 12 10 AM
After:
Screen Shot 2022-02-13 at 7 11 26 AM

### CODE REVIEW

**Review by babolivier — CHANGES_REQUESTED** [review]
After discussing this PR in [`#synapse-dev:matrix.org`](https://matrix.to/#/#synapse-dev:matrix.org), we've concluded that this is too specific/hacky to be accepted into Synapse. As much as possible, we would like to avoid having to rely on idioms and pseudo-conventions that we would have to change every time a big website changes its behaviour.

The compromise we're willing to make, though, is modifying the user-agent to mention "bot" in its name, so for example "Synapse (bot)", when fetching URL previews. While I still wish that we didn't have to do that (because it's at most a convention where we should be relying on standards), it should be generic enough and existing for long enough that we could accept it in Synapse.

Worth noting the solution of adding "bot" to the UA would fix the specific case of Twitter URL previews in an identical way to your proposed change:

```
$ curl -s -A "facebookexternalhit/1.1 (+http://www.facebook.com/externalhit_uatext.php)" "https://twitter.com/matrixdotorg/status/1492207278560268289" | grep -i 'og:'
    <meta  property="og:type" content="article">
    <meta  property="og:url" content="https://twitter.com/matrixdotorg/status/1492207278560268289">
    <meta  property="og:title" content="Matrix on Twitter">
    <meta  property="og:image" content="https://pbs.twimg.com/media/FLVWl6bXEAElabl.jpg:large">
    <meta  property="og:image:user_generated" content="true">
    <meta  property="og:image" content="https://pbs.twimg.com/media/FLVWttFWUAEUQ8v.png:large">
    <meta  property="og:image:user_generated" content="true">
    <meta  property="og:image" content="https://pbs.twimg.com/media/FLVWzSlXEAQwrqi.png:large">
    <meta  property="og:image:user_generated" content="true">
    <meta  property="og:description" content="“🗞️ TWIM is out!&#10;&#10;This week all major homeservers got an update, @onbeeper was released out of beta, @kdecommunity&#39;s NeoChat is now sharing-friendly, @element_hq gets experimental support for threads… and of course much more in the latest TWIM issue!&#10;&#10;https://t.co/ZLD9JaOTLG”">
    <meta  property="og:site_name" content="Twitter">
$ curl -s -A "Synapse (bot)" "https://twitter.com/matrixdotorg/status/1492207278560268289" | grep -i 'og:'
    <meta  property="og:type" content="article">
    <meta  property="og:url" content="https://twitter.com/matrixdotorg/status/1492207278560268289">
    <meta  property="og:title" content="Matrix on Twitter">
    <meta  property="og:image" content="https://pbs.twimg.com/media/FLVWl6bXEAElabl.jpg:large">
    <meta  property="og:image:user_generated" content="true">
    <meta  property="og:image" content="https://pbs.twimg.com/media/FLVWttFWUAEUQ8v.png:large">
    <meta  property="og:image:user_generated" content="true">
    <meta  property="og:image" content="https://pbs.twimg.com/media/FLVWzSlXEAQwrqi.png:large">
    <meta  property="og:image:user_generated" content="true">
    <meta  property="og:description" content="“🗞️ TWIM is out!&#10;&#10;This week all major homeservers got an update, @onbeeper was released out of beta, @kdecommunity&#39;s NeoChat is now sharing-friendly, @element_hq gets experimental support for threads… and of course much more in the latest TWIM issue!&#10;&#10;https://t.co/ZLD9JaOTLG”">
    <meta  property="og:site_name" content="Twitter">
```

**Comment by AndrewRyanChama** [comment]
Thanks for the input.  I updated the user agent string here to the one proposed

File: synapse/rest/media/v1/preview_url_resource.py [hunk 1]
```diff
@@ -402,7 +403,10 @@ async def _download_url(self, url: str, output_stream: BinaryIO) -> DownloadResu
                 url,
                 output_stream=output_stream,
                 max_size=self.max_spider_size,
-                headers={"Accept-Language": self.url_preview_accept_language},
+                headers={
+                    b"Accept-Language": self.url_preview_accept_language,
+                    b"User-Agent": ["Synapse (bot)"],
```
Comment (babolivier) [inline synapse/rest/media/v1/preview_url_resource.py:408]:
```suggestion
                    "Accept-Language": self.url_preview_accept_language,
                    "User-Agent": ["Synapse (bot)"],
```

For consistency.

File: synapse/res/providers.json [hunk 2]
```diff
@@ -5,13 +5,11 @@
         "endpoints": [
             {
                 "schemes": [
-                    "https://twitter.com/*/status/*",
-                    "https://*.twitter.com/*/status/*",
```
Comment (babolivier) [inline synapse/res/providers.json:9]: Is this removal still necessary? If so, why?

**Review by babolivier — CHANGES_REQUESTED** [review]

File: synapse/rest/media/v1/preview_url_resource.py (hunk 1 above)
Comment (babolivier) [reply synapse/rest/media/v1/preview_url_resource.py:408]: Oh yes, very good point.

File: synapse/res/providers.json (hunk 2 above)
Comment (clokep) [reply synapse/res/providers.json:9]: It (+ the changes to prefer oEmbed vs. HTML Open Graph) seem to be an unrelated change to what is proposed in the description.

File: synapse/rest/media/v1/preview_url_resource.py (hunk 1 above)
Comment (AndrewRyanChama) [reply synapse/rest/media/v1/preview_url_resource.py:408]: in SimpleHttpClient the headers are all specified as bytes strings.  If we pass this map with a regular string key, python will not overwrite the default user-agent key and put both 'User-Agent' and b'User-agent' both in the headers map.  Since SimpleHttpClient uses bytes strings for all headers throughtout client.py, I believe that this is the correct behavior.

I will update the agent string as suggested

File: synapse/res/providers.json (hunk 2 above)
Comment (AndrewRyanChama) [reply synapse/res/providers.json:9]: Sites listed in providers.json will use the specified oEmbed url and completely bypass opengraph parsing of the page being previewed.

Since this change is to fix previously broken opengraph previews, we should likewise remove these oEmbed overrides for the pages that we fixed.

Comment (AndrewRyanChama) [reply synapse/res/providers.json:9]: @clokep  removed the prefer change from this PR since that is not related.

**Comment by anoadragon453** [comment]
@AndrewRyanChama heads up that synapse maintainers prefer that PR branches aren't rebased after each update, in order to make it easier to see what changed between reviews.

We squash and merge PRs at the end of the review cycle, so it'll all end up in one commit before hitting `develop` anyhow.

File: synapse/res/providers.json (hunk 2 above)
Comment (babolivier) [reply synapse/res/providers.json:9]: > @clokep removed the prefer change from this PR since that is not related.

Sorry, what did you remove? I can still see this change in the PR's diff.

File: synapse/rest/media/v1/preview_url_resource.py [hunk 3]
```diff
@@ -402,7 +402,12 @@ async def _download_url(self, url: str, output_stream: BinaryIO) -> DownloadResu
                 url,
                 output_stream=output_stream,
                 max_size=self.max_spider_size,
-                headers={"Accept-Language": self.url_preview_accept_language},
+                headers={
+                    b"Accept-Language": self.url_preview_accept_language,
+                    b"User-Agent": [
+                        "Synapse (bot; +https://github.com/matrix-org/synapse)"
+                    ],
```
Comment (clokep) [inline synapse/rest/media/v1/preview_url_resource.py:409]: This does something a little odd where it stomps over the header that gets configured from the homeserver config. I feel like this at the very least needs a comment.

Comment (squahtx) [reply synapse/rest/media/v1/preview_url_resource.py:409]: We should also note that we're omitting the Synapse version for security reasons (#4376), in case anyone's tempted to add it back in.

File: synapse/res/providers.json (hunk 2 above)
Comment (AndrewRyanChama) [reply synapse/res/providers.json:9]: Removed the change on line 330 https://github.com/matrix-org/synapse/pull/11985/files#diff-bab0f9e113994dc8e6d6dedc4e2d3ce8b5ff6e059dc181ccf4db4c3358d60b29R330

I believe the providers change should be part of this PR since the purpose is to fix the opengraph fetching.  This providers list will bypass the change completely.

Comment (AndrewRyanChama) [reply synapse/res/providers.json:9]: I removed the change on line 330 from this PR.  This PR now changes only the user agent and providers.json, and is unrelated to #11669 now.

The change to providers.json is necessary specifically so twitter urls will use opengraph instead of bypassing the logic.

File: synapse/rest/media/v1/preview_url_resource.py [hunk 4]
```diff
@@ -402,7 +402,15 @@ async def _download_url(self, url: str, output_stream: BinaryIO) -> DownloadResu
                 url,
                 output_stream=output_stream,
                 max_size=self.max_spider_size,
-                headers={"Accept-Language": self.url_preview_accept_language},
+                headers={
+                    b"Accept-Language": self.url_preview_accept_language,
+                    # Use a custom user agent for the preview because some sites will only return
+                    # openGraph metadata to crawler user agents. We specifically omit the version
+                    # string to avoid leaking the this information.
```
Comment (clokep) [inline synapse/rest/media/v1/preview_url_resource.py:409]:
```suggestion
                    # Use a custom user agent for the preview because some sites will only return
                    # Open Graph metadata to crawler user agents. Omit the Synapse version
                    # string to avoid leaking information.
```

File: synapse/res/providers.json [hunk 5]
```diff
@@ -5,13 +5,11 @@
...
                     "https://twitter.com/*/moments/*",
                     "https://*.twitter.com/*/moments/*"
```
Comment (clokep) [inline synapse/res/providers.json:9]: Weirdly these URLs don't seem to  be used by Twitter anymore? We can leave them for now, but it is a bit annoying they changed their URL scheme at some point...

File: changelog.d/11985.misc [hunk 6]
```diff
@@ -0,0 +1 @@
+Use a bot User-Agent for URL preview queries.
```
Comment (clokep) [inline changelog.d/11985.misc:1]: Thinking through this more, I think we should make this a `.feature` file which says "Fetch images when previewing Twitter URLs."

You can also credit yourself if you would like "Contributed by (your name or @AndrewRyanChama)."

**Review by clokep — COMMENTED** [review]
Tested a bunch of URLs with a janky test rig and this seems to have the following changes:

* Tweets now have images (either from the tweet itself or from the profile).
* Twitter "events" now work instead of giving a JavaScript error.
* Some reddit links now include the Reddit logo as an image and include a little bit more description (the first sentence of the post instead of just the vote counts).

Other tested sites had no change (YouTube, GitHub, Flickr, Tenor, Imgur, etc.)

Just a little bit of bookkeeping and I think we can merge this!

**Review by clokep — APPROVED** [review]

**Comment by clokep** [comment]
Thanks for working through this with us! 🎉

**Comment by clokep** [comment]
For cross-linking, there was previously #7643.
